File: slices/scopes/scopes-rust/src/scope/unknown.rs
//! The fallback scope kind: anything the grammar doesn't recognize.

use std::fmt;

/// A scope string the grammar didn't recognize, preserved verbatim so it
/// round-trips unchanged. This is the total-parse fallback — once a string
/// reaches here it's neither a [`KnownScope`](crate::KnownScope), a FHIR
/// resource scope, nor a Wildflower resource scope.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct UnknownScope(String);

impl UnknownScope {
    pub(in crate::scope) fn new(s: &str) -> Self {
        UnknownScope(s.to_string())
    }
}

impl fmt::Display for UnknownScope {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

File: slices/scopes/scopes-core/src/domain/unknown.ts
/**
 * The fallback scope kind: anything the grammar doesn't recognize, preserved
 * verbatim so it round-trips unchanged. Mirrors `scopes-rust`'s `UnknownScope`
 * (`scope/unknown.rs`) — the total-parse fallback.
 *
 * Namespace module (`import { UnknownScope } from 'scopes-core'`): the value is
 * {@link UnknownScope}, with `UnknownScope.make`, `UnknownScope.scopeSerialize`.
 */

import type { ScopeSerializer } from '../behaviour/index.ts'

/** A scope string the grammar didn't recognize, kept verbatim. */
type UnknownScope = { readonly kind: 'unknown'; readonly raw: string }

/** Wrap a raw string as an unknown scope. */
const make = (raw: string): UnknownScope => ({ kind: 'unknown', raw })

/** Render an unknown scope — the raw string, unchanged. */
const scopeSerialize: ScopeSerializer<UnknownScope>['scopeSerialize'] = (scope) => scope.raw

export { type UnknownScope, make, scopeSerialize }
